sup3r/models/base.py

```diff
@@ -1052,3 +897,279 @@ def train(
                 break
 
         batch_handler.stop()
+
+    def calc_val_loss(self, batch_handler, weight_gen_advers):
+        """Calculate the validation loss at the current state of model training
+
+        Parameters
+        ----------
+        batch_handler : sup3r.preprocessing.BatchHandler
+            BatchHandler object to iterate through
+        weight_gen_advers : float
+            Weight factor for the adversarial loss component of the generator
+            vs. the discriminator.
+
+        Returns
+        -------
+        loss_details : dict
+            Running mean for validation loss details
+        """
+        logger.debug('Starting end-of-epoch validation loss calculation...')
+        for batch in batch_handler.val_data:
+            _, v_loss_details, _, _ = self._get_hr_exo_and_loss(
+                batch.low_res,
+                batch.high_res,
+                weight_gen_advers=weight_gen_advers,
+                train_gen=False,
+                train_disc=False,
+            )
+            self._val_record = self.update_loss_details(
+                self._val_record,
+                v_loss_details,
+                len(batch_handler.val_data),
+                prefix='val_',
+            )
+        return self._val_record.mean(axis=0)
+
+    def _get_batch_loss_details(
```

Comment:
There's a lot more going on here than just getting the loss details! This is running a full gradient descent step including updating model parameters. Function name and docstring are misleading. 

I'm on the fence on this one, i'm not convinced we need to split this out into its own function. for similar reasons (it's only called once, would we ever call this outside of a training loop?). There are quite a few lines but it's not that complicated and neither is the parent function.

Reply:
So now we have `_run_gradient_descent` and `run_gradient_descent` and they both take different args and output different things? I don't love that haha. If you simply must have this be a separate function, what about `_run_gradient_descent` -> `_train_batch` and then maybe also consider if `train_epoch` should be hidden `_train_epoch` to match.